rust/src/dcerpc/dcerpc.rs

```diff
@@ -0,0 +1,2600 @@
+                -1
+            }
+            Err(_) => {
+                // Error, probably malformed data.
+                SCLogDebug!("An error occurred while parsing DCERPC BINDACK");
+                -1
+            }
+        }
+    }
+
+    pub fn handle_stub_data(&mut self, input: &[u8], input_len: u16) -> u16 {
+        let mut retval: u16 = 0;
+        let hdrpfcflags = match self.get_hdr_pfcflags() {
+            Some(x) => x,
+            None => 0,
+        };
+        let rpc_vers = match self.get_hdr_rpc_vers() {
+            Some(x) => x,
+            None => 0,
+        };
+        let padleft = self.padleft;
+        // Update the stub params based on the packaet type
+        match self.get_hdr_type() {
+            Some(x) => match x {
+                DCERPC_TYPE_REQUEST => {
+                    if let Some(ref mut req) = self.request {
+                        retval = evaluate_stub_params(
+                            input,
+                            input_len,
+                            hdrpfcflags,
+                            padleft,
+                            &mut req.stub_data_buffer,
+                            &mut req.stub_data_buffer_len,
+                            &mut req.stub_data_buffer_reset,
+                            rpc_vers,
+                        );
+                    }
+                }
+                DCERPC_TYPE_RESPONSE => {
+                    if let Some(ref mut resp) = self.response {
+                        retval = evaluate_stub_params(
+                            input,
+                            input_len,
+                            hdrpfcflags,
+                            padleft,
+                            &mut resp.stub_data_buffer,
+                            &mut resp.stub_data_buffer_len,
+                            &mut resp.stub_data_buffer_reset,
+                            rpc_vers,
+                        );
+                    }
+                }
+                _ => {
+                    SCLogDebug!("Unrecognized packet type");
+                    return 0;
+                }
+            },
+            None => {
+                return 0;
+            }
+        }
+        // Update the remaining fragment length
+        self.padleft -= retval;
+
+        retval
+    }
+
+    pub fn handle_common_stub(&mut self, input: &[u8], bytes_consumed: i32, dir: u8) -> i32 {
+        let fraglen = match self.get_hdr_fraglen() {
+            Some(x) => x,
+            None => {
+                return -1;
+            }
+        };
+        self.padleft = fraglen - 16 - bytes_consumed as u16; // bytes processed so far = header length
+        let mut input_left = input.len() as i32 - bytes_consumed;
+        let mut parsed: i32 = bytes_consumed; // parsed bytes
+        while input_left > 0 && parsed < fraglen as i32 {
+            let retval = self.handle_stub_data(&input[parsed as usize..], input_left as u16);
```

Comment:
is this cast to u16 down from u32 of input_left safe? Is it possible that it is > 64k?

Reply:
DCE 1.1 defines fraglen as u16 (https://pubs.opengroup.org/onlinepubs/9629399/chap12.htm), I think I was keeping this variable as i32 because in this while loop things are subtracted from it so there's a possibility of things going in negative (not in ideal scenario though). Am I thinking wrong?

Reply:
So in what case can you see it going negative?

Reply:
None. Didn't check it properly. Thanks.